src/extensions/markdown/Lists/plugins/MergeListsPlugin.ts

```diff
@@ -27,12 +27,35 @@ function mergeAdjacentNodesWithSameType(
     tr: Transaction,
     nodes: ReturnType<typeof findChildren>,
 ): void {
-    for (let i = nodes.length - 1; i > 0; i--) {
-        const prev = nodes[i - 1];
-        const next = nodes[i];
+    const lowerNodes = [];
+
+    // сache node depths to avoid multiple resolve calls
+    const nodeDepths = nodes.map((node) => tr.doc.resolve(node.pos).depth);
 
-        if (prev.node.type === next.node.type && prev.pos + prev.node.nodeSize === next.pos) {
-            tr.join(next.pos);
+    // еraverse the nodes array from the end for optimization purposes
+    for (let i = nodes.length - 1; i > 0; i--) {
+        const upperNode = nodes[i - 1];
+        const currentNode = nodes[i];
+
+        const upperNodeDepth = nodeDepths[i - 1];
+        const currentNodeDepth = nodeDepths[i];
+
+        if (upperNodeDepth > currentNodeDepth) {
+            // save the current node for potential later merging
+            // when we encounter nodes at the same level
+            lowerNodes.push(currentNode);
+        } else {
+            // compare nodes only at the same level of nesting
+            const lowerNode = upperNodeDepth === currentNodeDepth ? currentNode : lowerNodes.pop();
+
+            if (
+                lowerNode &&
+                upperNode.node.type === lowerNode.node.type &&
+                upperNode.pos + upperNode.node.nodeSize === lowerNode.pos &&
+                upperNode.node.attrs.bullet === lowerNode.node.attrs.bullet
```

Comment:
Note: this check ensures that the join() function works correctly.

Also, such work corresponds to the work of the diplodoc:
<img width="1000" alt="diplodoc lists" src="https://github.com/user-attachments/assets/3d144851-599f-424f-a510-f5b473eb31ff">